Sept updates (#56)
YCloudYUSA/yusaopeny_docs: content/en/docs/development/drupal-10-update/_index.md

- If errors occur, review the conflicts, check out the [known issues](https://www.drupal.org/docs/updating-drupal/updating-drupal-core-via-composer#known-issues), and attempt to resolve them.
- Re-run the previous steps until they complete successfully
- Run `drush updb`, review the updates, and run them.

## Smoke tests

We recommend reviewing critical functionality after the update to ensure any custom functionality still works.

## Troubleshooting

### Composer issues

Composer can be ... tricky. To resolve composer conflicts:

- If specific modules conflict, try requiring them directly to get more information about the conflict.
- Make good use of [composer why](https://getcomposer.org/doc/03-cli.md#depends-why) and [composer why-not](https://getcomposer.org/doc/03-cli.md#prohibits-why-not)

### Update hook conflicts

If you run into an error like this:

> \>  [notice] Update started: y_lb_update_9011
>
> \> [error]  Configuration <em class="placeholder">core.entity_view_display.node.lb_event.featured</em> depends on the <em class="placeholder">core.entity_view_mode.node.featured</em> configuration that will not exist after import.

you may be able to resolve it yourself:

> 1. Search the codebase for the missing config.

Reviewer: can you add an example of how to search?  i.e. `grep -rI lb_event .` from the `docroot` parent folder

> 2. Add an update hook to your own custom module [like this example](https://github.com/YCloudYUSA/yusaopeny/pull/108/files#diff-2b10287d954ae6b7c36a1af05970dc0f2ee602047b2215d60d9052a15a8819b5R1233-R1240):

Reviewer: which file does this code go in?

    ```php
    function mymodule_update_9000() {
        $path = \Drupal::service('extension.list.module')->getPath('ws_event') . '/config/optional';
        /** @var \Drupal\config_import\ConfigImporterService $config_importer */
        $config_importer = \Drupal::service('config_import.importer');
        $config_importer->setDirectory($path);
        $config_importer->importConfigs([
            'core.entity_view_mode.node.featured',
        ]);
    }
    ```

> 3. Use [hook_update_dependencies](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_update_dependencies/10.0.x) to ensure this new update runs before the failing one.

Reviewer: which file does this code go in?

## Resources

- [Upgrading from Drupal 9 to Drupal 10 (drupal.org)](https://www.drupal.org/docs/upgrading-drupal/upgrading-from-drupal-8-or-later/upgrading-from-drupal-9-to-drupal-10)
- [Deprecated and obsolete extensions (drupal.org)](https://www.drupal.org/docs/core-modules-and-themes/deprecated-and-obsolete)
- [mglaman/composer-drupal-lenient](https://github.com/mglaman/composer-drupal-lenient) - can help install modules that do not yet have official Drupal 10 support.

## Review important versions

Step through the distribution's [important versions](../important-versions-for-upgrade-path/#important-versions) until you reach `9.2.13.0`. You should be running the latest Drupal 9.5.x before you begin the upgrade to 10.

## Pre-checks

1. CKEditor
    1. If any custom/contrib modules are used, CKE5 should likely be done AFTER your D10 upgrade
    2. Contrib checks will NOT be found in the next step, be sure to check these manually
2. Dependency cleanup
    1. Modules not installed, but in composer.json should be cleaned up to prevent unwanted dependency issues in trying to update.
3. Admin theme

>     1. If your website uses a [deprecated admin theme](https://www.drupal.org/docs/core-modules-and-themes/deprecated-and-obsolete#s-recommendations-for-deprecated-themes), you should migrate to Claro theme and test the admin experience. If necessary you can keep the deprecated theme as a contrib package but that is not recommended.

Reviewer 2: ```suggestion
    1. If your website uses a [deprecated admin theme](https://www.drupal.org/docs/core-modules-and-themes/deprecated-and-obsolete#s-recommendations-for-deprecated-themes), you should migrate to Claro theme and test the admin experience. If necessary you can keep the deprecated theme as a contrib package but that is not recommended and won't be supported by this distribution.
```

Author: ```suggestion
    1. If your website uses a [deprecated admin theme](https://www.drupal.org/docs/core-modules-and-themes/deprecated-and-obsolete#s-recommendations-for-deprecated-themes), you should migrate to the Claro theme and test the admin experience. If necessary you can keep the deprecated theme as a contrib package but that is not recommended and won't be supported by the distribution.
```

yes and one more tweak i saw.

    ```php
   // This goes in mymodule.install as the next update hook.
   // Increment the number accordingly.
    function mymodule_update_9000() {
        $path = \Drupal::service('extension.list.module')->getPath('ws_event') . '/config/optional';
        /** @var \Drupal\config_import\ConfigImporterService $config_importer */
        $config_importer = \Drupal::service('config_import.importer');
        $config_importer->setDirectory($path);
        $config_importer->importConfigs([
            'core.entity_view_mode.node.featured',
        ]);
    }
    ```

2. Use [hook_update_dependencies](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_update_dependencies/10.0.x) to ensure this new update runs before the failing one.

    ```php
   // This also goes in mymodule.install.
    function mymodule_update_dependencies() {
      $dependencies['y_lb'][9011] = [
        'mymodule' => 9000,
      ];
    }
    ```

3. Re-run `drush updb`.

> 4. If you run into other missing config, add them to the list to be imported in `update` hook and re-run `updb`.

Reviewer 2: ```suggestion
4. If you run into other missing configs, add them to the list to be imported in `update` hook and re-run `updb`.
5. Consider to backport your customisation which led to the challenge of doing this upgrade in order for it to be covered and tested by  distribution developers.
```
